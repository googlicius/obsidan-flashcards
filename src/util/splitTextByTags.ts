export function splitTextByTags(text: string, tags: string[]): string[][] {
    // Convert tags array to regex pattern
    const tagPattern = tags.map(tag => tag.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')).join('|');
    
    // Create regex that matches tags only when they're on their own line
    const regex = new RegExp(`^[\\t ]*(${tagPattern})[\\t ]*$`, 'gm');
    
    // Find all matches with their positions
    const matches = [...text.matchAll(regex)];
    
    // If no matches found, return original text in array
    if (matches.length === 0) {
        return [['', text]];
    }

    const results: string[][] = [];
    let lastIndex = 0;

    for (let i = 0; i < matches.length; i++) {
        const match = matches[i];

        // If this is not the first match and there's content between matches
        if (i > 0) {
            const prevMatch = matches[i-1];
            const currentSection = text.slice(lastIndex, match.index ?? 0).trimEnd();
            results.push([prevMatch[1], currentSection]);
        }
        
        // Use nullish coalescing operator to handle potential undefined
        lastIndex = match.index ?? 0;
        
        // If this is the last match, add the remaining text
        if (i === matches.length - 1) {
            const finalSection = text.slice(lastIndex).trimEnd();
            results.push([match[1], finalSection]);
        }
    }

    return results;
}
