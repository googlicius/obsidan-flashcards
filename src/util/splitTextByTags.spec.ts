import { splitTextByTags } from './splitTextByTags';

describe('splitTextByTags', () => {
    it('should split the text by 1 tag', () => {
        const text = `

        #flashcard

        #tag3

        # tag1 Test 3 // Not split this because it's mixing with text
        Test 4

        #tag2 

        AAAAAA :: BBBBBB
                    `;
        
                const tags = ['#flashcard'];
                const results = splitTextByTags(text, tags);
        
                expect(results).toMatchSnapshot();
    });

	it('should split the text by 2 given tags', () => {
		const text = `
#tag1

#tag3

# tag1 Test 3 // Not split this because it's mixing with text
Test 4

#tag2 

AAAAAA :: BBBBBB
            `;

		const tags = ['#tag1', '#tag2'];
		const results = splitTextByTags(text, tags);

		expect(results).toMatchSnapshot();
	});
});
