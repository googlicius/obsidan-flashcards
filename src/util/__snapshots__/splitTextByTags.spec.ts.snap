// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`splitTextByTags should split the text by 1 tag 1`] = `
[
  [
    "#flashcard",
    "        #flashcard

        #tag3

        # tag1 Test 3 // Not split this because it's mixing with text
        Test 4

        #tag2 

        AAAAAA :: BBBBBB",
  ],
]
`;

exports[`splitTextByTags should split the text by 2 given tags 1`] = `
[
  [
    "#tag1",
    "#tag1

#tag3

# tag1 Test 3 // Not split this because it's mixing with text
Test 4",
  ],
  [
    "#tag2",
    "#tag2 

AAAAAA :: BBBBBB",
  ],
]
`;
